### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by jpiontek**

### CODE REVIEW

**Comment by gathreya** [comment]
Awesomeness :)

File: config/local.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+const BINPATH = './node_modules/nightwatch/bin/'
```
Comment (gathreya) [inline config/local.js:1]: Do we need to include these jars given it will be downloaded anyway? Same wrt chromedriver

File: tests/award.js [hunk 2]
```diff
@@ -1,383 +1,384 @@
 const assert = require('assert');
 
 module.exports = {
-  '@disabled': true,
-  // '@tags': ['authorization', 'admin'],
-  before: function (browser) {
-  },
+    '@disabled': true,
+    // '@tags': ['authorization', 'admin'],
```
Comment (gathreya) [inline tests/award.js]: Can we enable this test?

File: tests/award.js [hunk 3]
```diff
@@ -1,383 +1,384 @@
...
-            .clearValue('input[id="document.budget.budgetPeriods[0].directCostLimit"]')
-            .setValue('input[id="document.budget.budgetPeriods[0].directCostLimit"]', '25000.00')
-            .clearValue('input[id="document.budget.budgetPeriods[1].directCostLimit"]')
-            .setValue('input[id="document.budget.budgetPeriods[1].directCostLimit"]', '25000.00')
-            .clearValue('input[id="document.budget.budgetPeriods[2].directCostLimit"]')
-            .setValue('input[id="document.budget.budgetPeriods[2].directCostLimit"]', '25000.00')
-            .clearValue('input[id="document.budget.budgetPeriods[3].directCostLimit"]')
-            .setValue('input[id="document.budget.budgetPeriods[3].directCostLimit"]', '25000.00')
-
-            .click('input[name="methodToCall.headerTab.headerDispatch.save.navigateTo.expenses"]')
```
Comment (gathreya) [inline tests/award.js:180]: What are these changes?

File: tests/proposalDevelopment.js [hunk 4]
```diff
@@ -1,42 +1,42 @@
 module.exports = {
     '@disabled': true,
-
```
Comment (gathreya) [inline tests/proposalDevelopment.js:3]: When you enable other tests, leave this one disabled, it is not complete.

File: tests/proposalDevelopment.js [hunk 5]
```diff
@@ -1,42 +1,42 @@
...
+            .setValue('textArea[name="document.developmentProposal.title"]', 'Test')
+            .click('button[id=ufuknm4]')
+            .pause(1000)
 
-    //.end();
+        //.end();
     }
-};
\ No newline at end of file
+};
```
Comment (gathreya) [inline tests/proposalDevelopment.js:39]: I am guessing all of the test changes are just whitespace changes? Bad diff :(

File: package.json [hunk 6]
```diff
@@ -20,8 +20,5 @@
     "nightwatch": "0.9.12",
     "selenium-download": "^2.0.10",
     "superagent": "3.5.2"
-  },
-  "devDependencies": {
-    "browserstack-local": "1.3.0"
   }
```
Comment (gathreya) [inline package.json:23]: Why did you remove this? This is what connects the local tests to browserstack right? Sorry, don't remember.

File: package.json [hunk 7]
```diff
@@ -7,7 +7,7 @@
     "test": "tests"
   },
   "scripts": {
-    "test": "node ./nightwatch-runner.js -c ./nightwatch-config.js",
+    "test": "node ./nightwatch-runner.js --config ./config/browserstack.js",
     "test-local": "node_modules/.bin/nightwatch --config ./config/local.js",
     "postinstall": "node ./config/local.js"
   },
```
Comment (gathreya) [inline package.json]: Nice, love it. Won't have to maintain tests in 2 places anymore :)

File: config/browserstack.js [hunk 8]
```diff
@@ -7,24 +7,86 @@
...
-  nightwatch: {
-    baseUrl: getenv.string(
-      'NIGHTWATCH_BASE_URL',
-      'https://res-tst1.kuali.co'
-    ),
-    debug: getenv.bool('NIGHTWATCH_DEBUG', false),
-    local: getenv.bool('NIGHTWATCH_LOCAL', false),
-    runInParallel: getenv.bool('NIGHTWATCH_RUN_IN_PARALLEL', false)
+
+  test_workers: getenv.bool('NIGHTWATCH_RUN_IN_PARALLEL', false),
```
Comment (gathreya) [inline config/browserstack.js]: We should probably enable this so we can save time.

File: config/browserstack.js [hunk 9]
```diff
@@ -7,24 +7,86 @@
...
+    //baseUrl: getenv.string(
+      //'NIGHTWATCH_BASE_URL',
+      //'https://res-tst1.kuali.co'
+    //),
+    //debug: getenv.bool('NIGHTWATCH_DEBUG', false),
+    //local: getenv.bool('NIGHTWATCH_LOCAL', false),
+    //runInParallel: getenv.bool('NIGHTWATCH_RUN_IN_PARALLEL', false)
+  //}
+//}
+
```
Comment (gathreya) [inline config/browserstack.js]: Remove commented out code.

File: tests/award.js [hunk 10]
```diff
@@ -377,7 +376,6 @@ module.exports = {
                             .end();
                         done();
                     })
-                done();
```
Comment (gathreya) [inline tests/award.js]: Huh, does this finish without the done() ?

File: tests/award.js [hunk 11]
```diff
@@ -377,6 +377,6 @@ module.exports = {
                         done();
                     })
             })
-            .end();
+            //.end();
     }
```
Comment (gathreya) [inline tests/award.js]: Same comment here, does the test finish without the end() or does the session hang?

File: config/browserstack.js (hunk 9 above)
Comment (gathreya) [reply config/browserstack.js]: So.... that previous comment was done for another commit. Odd how github shows that in this commit.

**Comment by gathreya** [comment]
Love it overall, but few minor issues / questions. Need to modify the jenkins job as needed?

File: package.json (hunk 6 above)
Comment (jpiontek) [reply package.json:23]: This is for using the browserstack local plugin, to run them locally. However, the changes make it so that nightwatch uses selenium and chrome directly and you don't have to use browserstack.

File: config/browserstack.js (hunk 8 above)
Comment (jpiontek) [reply config/browserstack.js]: yeah, for now though it makes it easier to troubleshoot things. This should speed things up, though.

File: tests/proposalDevelopment.js (hunk 5 above)
Comment (jpiontek) [reply tests/proposalDevelopment.js:39]: yeah I tried to fix some whitespace/indentation issues...

File: tests/award.js (hunk 3 above)
Comment (jpiontek) [inline tests/award.js]: These shouldn't be removed. Must have committed some garbage. I'll fix this before merge.

File: config/local.js (hunk 1 above)
Comment (jpiontek) [reply config/local.js:1]: Yes, the npm module will just download them and put them in the correct place - still need to point nightwatch to where they are.

File: tests/award.js (hunk 11 above)
Comment (jpiontek) [reply tests/award.js]: The tests calls end twice, which destroys the session on the driver. The end() call in the perform block just above this should properly end the session after the async stuff is done. Otherwise you'll destroy the session before the async stuff finishes.

File: tests/award.js (hunk 10 above)
Comment (jpiontek) [inline tests/award.js]: there's an issue with this test - if you look there's 2 nested .perform methods being called and they both have a "done" param. So the inner done is shadowing the outer done which leads to some weird behavior. Plus, I think all the async stuff in .perform methods isn't even completing before the test finishes.

File: tests/award.js (hunk 3 above)
Comment (jpiontek) [reply tests/award.js:180]: Spoke too soon, I think this is just the diff getting crazy when I fixed the indentation in the file.

File: tests/award.js (hunk 10 above)
Comment (jpiontek) [reply tests/award.js]: I think we need to revisit this test Since the .perform functions are all async in this case, things are completing before the test is finished when done() is called.

**Comment by jpiontek** [comment]
I actually don't think anything will have to change on the jenkins job. It should still execute the same script via "npm test". The only thing would be to remove the node modules and re-run npm install so that the selenium-download module will run and the postinstall script will move the driver and .jar file where they need to be.

File: package.json (hunk 6 above)
Comment (gathreya) [reply package.json:23]: OK but we CAN use browserstack when we want to right?

File: tests/award.js (hunk 10 above)
Comment (gathreya) [reply tests/award.js]: That's interesting. I know for sure that the test gets those values from the perform function before the test completes though because the test would not be able to proceed all the way to the end without some of the values calculated in those async functions. Without the done, the test was just hanging even after completion.

**Review by gathreya — APPROVED** [review]
